walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -11,13 +11,125 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
+ 
 package com.google.sps;
-
+ 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
-
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+ 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // create a full day of free timeslots
+    List<TimeRange> fullDay = new ArrayList<TimeRange>();
+    fullDay.add(TimeRange.WHOLE_DAY);
+    
+    // cannot modify arraylist in an enhanced for loop, must create separate arraylists
+    List<TimeRange> freeTimes = new ArrayList<TimeRange>();
+    List<TimeRange> blockedTimes = new ArrayList<TimeRange>();
+
+    TimeRange blocked;
+    int start, end;
+    Collection<String> attendees = request.getAttendees();
+    Set<String> eventAttendees;
+
+    for(Event event: events){
+        eventAttendees = event.getAttendees();
+        for(String attendee:eventAttendees){
+            if(attendees.contains(attendee)){
+              blocked = event.getWhen();
+              start = blocked.start();
+              end = blocked.end();
+               
+              for(TimeRange timeslot: fullDay){
+                if(blocked.overlaps(timeslot)){
+                    // case one: event is fully enclosed within a free timeslot
+                    if(start > timeslot.start() && end < timeslot.end()){
+                        freeTimes.add(TimeRange.fromStartEnd(timeslot.start(), start,false));
+                        freeTimes.add(TimeRange.fromStartEnd(end,timeslot.end(),false));
+                    }
+                    // case two: event overlaps end of free timeslot
+                    else if(start >= timeslot.start() && end >= timeslot.end()){
+                      if(start != timeslot.start()){
+                      freeTimes.add(TimeRange.fromStartEnd(timeslot.start(), start,false));
+                      }
+                    }
+                    // case three: event overlaps start of free timeslot
+                    else if(start <= timeslot.start() && end >= timeslot.start()){
+                        if(end != timeslot.end()){
+                            freeTimes.add(TimeRange.fromStartEnd(end,timeslot.end(),false));
+                        }
+                    }
+                    blockedTimes.add(timeslot);
+                }
+              }
+              fullDay.addAll(freeTimes);
+              fullDay.removeAll(blockedTimes);
+              // must clear everytime so that blocked events don't get added back
+              freeTimes.clear();
+              blockedTimes.clear();
+              // add a break to break early if one attendee is present at event
+              break; 
```

Comment:
Given how far this break is from the initial loop to check if there is any overlap in attendees, it's hard to tell at a glance what this break is for or that the preceeding loop is working correctly/not being repeated.

It would be much clearer and easier to understand what is going on if you separate the logic that decides if events have overlapping attendees from the logic to decide if the event overlaps any existing free time slots.

Reply:
okay, i'll add some kind of flag.

Reply:
Have a look at [`Collections.disjoint`](https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#disjoint(java.util.Collection,%20java.util.Collection))

Reply:
do you want me to use this in place of break?

Reply:
Done.